File: mobile/android/app/src/main/java/com/learno/app/ui/viewmodel/AuthViewModel.kt
package com.learno.app.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.learno.app.LearnoApplication
import com.learno.app.data.local.PreferencesManager
import com.learno.app.data.model.LoginRequest
import com.learno.app.data.model.RegisterRequest
import com.learno.app.data.model.User
import com.learno.app.data.remote.ApiClient
import com.learno.app.data.remote.api.AuthApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class AuthViewModel : ViewModel() {
    private val authApi = ApiClient.createService(AuthApi::class.java)
    private val preferencesManager = PreferencesManager(LearnoApplication.instance)

    private val _isAuthenticated = MutableStateFlow(checkAuthStatus())
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated

    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage

    init {
        if (_isAuthenticated.value) {
            loadCurrentUser()
        }
    }

    private fun checkAuthStatus(): Boolean {
        return preferencesManager.getAuthToken() != null
    }

    fun login(email: String, password: String) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                val response = authApi.login(LoginRequest(email, password))
                if (response.isSuccessful && response.body()?.success == true) {
                    response.body()?.data?.let { authResponse ->
                        ApiClient.setAuthToken(authResponse.token)
                        _currentUser.value = authResponse.user
                        _isAuthenticated.value = true
                    }
                } else {
                    _errorMessage.value = response.body()?.error ?: "Login failed"
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Network error"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun register(
        email: String,
        password: String,
        username: String,
        nativeLanguage: String,
        targetLanguages: List<String>
    ) {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                val response = authApi.register(
                    RegisterRequest(email, password, username, nativeLanguage, targetLanguages)
                )
                if (response.isSuccessful && response.body()?.success == true) {
                    response.body()?.data?.let { authResponse ->
                        ApiClient.setAuthToken(authResponse.token)
                        _currentUser.value = authResponse.user
                        _isAuthenticated.value = true
                    }
                } else {
                    _errorMessage.value = response.body()?.error ?: "Registration failed"
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Network error"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun logout() {
        ApiClient.clearAuthToken()
        _isAuthenticated.value = false
        _currentUser.value = null
    }

    private fun loadCurrentUser() {
        viewModelScope.launch {
            try {
                val response = authApi.getCurrentUser()
                if (response.isSuccessful && response.body()?.success == true) {
                    _currentUser.value = response.body()?.data
                } else {
                    logout()
                }
            } catch (e: Exception) {
                logout()
            }
        }
    }
}

File: mobile/ios/ViewModels/AuthViewModel.swift
import Foundation
import Combine
import SwiftUI

class AuthViewModel: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentUser: User?
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let authService = AuthService.shared
    private var cancellables = Set<AnyCancellable>()

    init() {
        checkAuthStatus()
    }

    private func checkAuthStatus() {
        // Check if we have a stored auth token
        if UserDefaults.standard.string(forKey: "authToken") != nil {
            loadCurrentUser()
        }
    }

    func login(email: String, password: String) {
        isLoading = true
        errorMessage = nil

        authService.login(email: email, password: password)
            .receive(on: DispatchQueue.main)
            .sink { [weak self] completion in
                self?.isLoading = false
                if case .failure(let error) = completion {
                    self?.errorMessage = error.localizedDescription
                }
            } receiveValue: { [weak self] authResponse in
                self?.currentUser = authResponse.user
                self?.isAuthenticated = true
                self?.saveUser(authResponse.user)
            }
            .store(in: &cancellables)
    }

    func register(
        email: String,
        password: String,
        username: String,
        nativeLanguage: String,
        targetLanguages: [String]
    ) {
        isLoading = true
        errorMessage = nil

        authService.register(
            email: email,
            password: password,
            username: username,
            nativeLanguage: nativeLanguage,
            targetLanguages: targetLanguages
        )
        .receive(on: DispatchQueue.main)
        .sink { [weak self] completion in
            self?.isLoading = false
            if case .failure(let error) = completion {
                self?.errorMessage = error.localizedDescription
            }
        } receiveValue: { [weak self] authResponse in
            self?.currentUser = authResponse.user
            self?.isAuthenticated = true
            self?.saveUser(authResponse.user)
        }
        .store(in: &cancellables)
    }

    func logout() {
        authService.logout()
        isAuthenticated = false
        currentUser = nil
        UserDefaults.standard.removeObject(forKey: "currentUser")
    }

    private func loadCurrentUser() {
        authService.getCurrentUser()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] completion in
                if case .failure = completion {
                    self?.logout()
                }
            } receiveValue: { [weak self] user in
                self?.currentUser = user
                self?.isAuthenticated = true
                self?.saveUser(user)
            }
            .store(in: &cancellables)
    }

    private func saveUser(_ user: User) {
        if let encoded = try? JSONEncoder().encode(user) {
            UserDefaults.standard.set(encoded, forKey: "currentUser")
        }
    }
}
